drivers/hwmon/surface_fan.c

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Surface Fan driver for Surface System Aggregator Module. It provides access
+ * to the fan's rpm through the hwmon system.
+ *
+ * Copyright (C) 2023 Ivor Wanders <ivor@iwanders.net>
+ */
+
+#include <linux/hwmon.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/surface_aggregator/device.h>
+#include <linux/types.h>
+
+// The minimum speed for the fan when turned on by the controller. The onboard
+// controller uses this as minimum value before turning the fan on or off.
+#define SURFACE_FAN_MIN_SPEED 3000
+// The maximum speed, determined by observation and rounding up to the nearest
+// multiple of 500 to account for variation between individual fans.
+#define SURFACE_FAN_MAX_SPEED 7500
+
+// SSAM
+SSAM_DEFINE_SYNC_REQUEST_CL_R(__ssam_fan_rpm_get, __le16, {
+	.target_category = SSAM_SSH_TC_FAN,
+	.command_id      = 0x01,
+});
+
+// hwmon
+umode_t surface_fan_hwmon_is_visible(const void *drvdata,
+				     enum hwmon_sensor_types type, u32 attr,
+				     int channel)
+{
+	if (type != hwmon_fan)
+		return 0;
+
+	switch (attr) {
+	case hwmon_fan_input:
+	case hwmon_fan_label:
+	case hwmon_fan_min:
+	case hwmon_fan_max:
+		return 0444;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static int surface_fan_hwmon_read(struct device *dev,
+				  enum hwmon_sensor_types type, u32 attr,
+				  int channel, long *val)
+{
+	struct ssam_device *sdev = dev_get_drvdata(dev);
+	__le16 value;
+	int res;
+
+	if (type != hwmon_fan)
+		return 0;
```

Comment:
Should this be `-1`?

Reply:
:facepalm:  yep... it probably should be, I copied from https://github.com/linux-surface/kernel/pull/144#discussion_r1428907537 without thinking.

Reply:
Fixed with f2d2b5647d5cceeca32ca52a0f19a81ef9509b81

Reply:
Also based on a very quick look around, I think this should be `-EOPNOTSUPP`, not `-1`.

Reply:
:+1:   a8eefe7f3651b395154c217b40cb49ce257468ba incorporates this.